bundles/org.openhab.binding.livisismarthome/src/main/java/org/openhab/binding/livisismarthome/internal/handler/LivisiBridgeHandler.java

```diff
@@ -0,0 +1,936 @@
+ * @author Sven Strohschein - Renamed from Innogy to Livisi
+ */
+@NonNullByDefault
+public class LivisiBridgeHandler extends BaseBridgeHandler
+        implements AccessTokenRefreshListener, EventListener, DeviceStatusListener {
+
+    private final Logger logger = LoggerFactory.getLogger(LivisiBridgeHandler.class);
+    private final Gson gson = new Gson();
+    private final Object lock = new Object();
+    private final Map<String, DeviceStatusListener> deviceStatusListeners;
+    private final OAuthFactory oAuthFactory;
+    private final HttpClient httpClient;
+
+    private @Nullable LivisiClient client;
+    private @Nullable LivisiWebSocket webSocket;
+    private @Nullable DeviceStructureManager deviceStructMan;
+    private @Nullable String bridgeId;
+    private @Nullable ScheduledFuture<?> reInitJob;
+    private @Nullable ScheduledFuture<?> bridgeRefreshJob;
+    private @NonNullByDefault({}) LivisiBridgeConfiguration bridgeConfiguration;
+    private @Nullable OAuthClientService oAuthService;
+    private String configVersion = "";
+
+    /**
+     * Constructs a new {@link LivisiBridgeHandler}.
+     *
+     * @param bridge Bridge thing to be used by this handler
+     * @param oAuthFactory Factory class to get OAuth2 service
+     * @param httpClient httpclient instance
+     */
+    public LivisiBridgeHandler(final Bridge bridge, final OAuthFactory oAuthFactory, final HttpClient httpClient) {
+        super(bridge);
+        this.oAuthFactory = oAuthFactory;
+        this.httpClient = httpClient;
+        deviceStatusListeners = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void handleCommand(final ChannelUID channelUID, final Command command) {
+        // not needed
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singleton(LivisiDeviceDiscoveryService.class);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing LIVISI SmartHome BridgeHandler...");
+        this.bridgeConfiguration = getConfigAs(LivisiBridgeConfiguration.class);
+        initializeClient();
+    }
+
+    /**
+     * Initializes the services and LivisiClient.
+     */
+    private void initializeClient() {
+        String tokenURL = URLCreator.createTokenURL(bridgeConfiguration.host);
+        final OAuthClientService oAuthServiceNonNullable = oAuthFactory.createOAuthClientService(
+                thing.getUID().getAsString(), tokenURL, tokenURL, "clientId", "clientPass", null, true);
+        this.oAuthService = oAuthServiceNonNullable;
+        LivisiClient clientNonNullable = createClient(oAuthServiceNonNullable);
+        this.client = clientNonNullable;
+        deviceStructMan = new DeviceStructureManager(createFullDeviceManager(clientNonNullable));
+
+        oAuthServiceNonNullable.addAccessTokenRefreshListener(this);
+        try {
+            requestAccessToken();
```

Comment:
`initialize()` should return fast, but this seems like a blocking operation.

Reply:
but should it return before the things are initialized (data is required to get loaded to initialize the things)? After the initial thing initialization, a task is scheduled to refresh data (that isn't blocking).

Reply:
Yes, before leaving `initialize()` you need to set status. You can set it to `ThingStatus.UNKNOWN`. Then, when data is received for the first time, status can be set to `ThingStatus.ONLINE`.

Reply:
solved